lib/internal/test_runner/tap_lexer.js

```diff
@@ -0,0 +1,522 @@
+'use strict';
+
+const { SafeSet, MathMax, StringPrototypeIncludes } = primordials;
+const {
+  codes: { ERR_TAP_LEXER_ERROR },
+} = require('internal/errors');
+
+const kEOL = '';
+const kEOF = '';
+
+const TokenKind = {
+  EOF: 'EOF',
+  EOL: 'EOL',
+  NEWLINE: 'NewLine',
+  NUMERIC: 'Numeric',
+  LITERAL: 'Literal',
+  KEYWORD: 'Keyword',
+  WHITESPACE: 'Whitespace',
+  COMMENT: 'Comment',
+  DASH: 'Dash',
+  PLUS: 'Plus',
+  HASH: 'Hash',
+  ESCAPE: 'Escape',
+  UNKNOWN: 'Unknown',
+
+  // TAP tokens
+  TAP: 'TAPKeyword',
+  TAP_VERSION: 'VersionKeyword',
+  TAP_PLAN: 'PlanKeyword',
+  TAP_TEST_POINT: 'TestPointKeyword',
+  TAP_SUBTEST_POINT: 'SubTestPointKeyword',
+  TAP_TEST_OK: 'TestOkKeyword',
+  TAP_TEST_NOTOK: 'TestNotOkKeyword',
+  TAP_YAML_START: 'YamlStartKeyword',
+  TAP_YAML_END: 'YamlEndKeyword',
+  TAP_YAML_BLOCK: 'YamlKeyword',
+  TAP_PRAGMA: 'PragmaKeyword',
+  TAP_BAIL_OUT: 'BailOutKeyword',
+};
+
+class Token {
+  constructor({ kind, value, stream }) {
+    const valueLength = ('' + value).length;
+    this.kind = kind;
+    this.value = value;
+    this.location = {
+      line: stream.line,
+      column: MathMax(stream.column - valueLength + 1, 1), // 1 based
+      start: MathMax(stream.pos - valueLength, 0), // zero based
+      end: stream.pos - (value === '' ? 0 : 1), // zero based
+    };
+
+    // EOF is a special case
+    if (value === TokenKind.EOF) {
+      const eofPosition = stream.input.length + 1; // We consider EOF to be outside the stream
+      this.location.start = eofPosition;
+      this.location.end = eofPosition;
+      this.location.column = stream.column + 1; // 1 based
+    }
+  }
+}
+
+class InputStream {
+  constructor(input) {
+    this.input = input;
+    this.pos = 0;
+    this.column = 0;
+    this.line = 1;
+  }
+
+  eof() {
+    return this.peek() === undefined;
+  }
+
+  peek(offset = 0) {
+    return this.input[this.pos + offset];
+  }
+
+  next() {
+    const char = this.peek();
+    if (char === undefined) {
+      return undefined;
+    }
+
+    this.pos++;
+    this.column++;
+    if (char === '\n') {
+      this.line++;
+      this.column = 0;
+    }
+
+    return char;
+  }
+}
+
+class TapLexer {
+  static Keywords = new SafeSet([
+    'TAP',
+    'version',
+    'ok',
+    'not',
+    '...',
+    '---',
+    '..',
+    'pragma',
+    '-',
+    '+',
+
+    // NOTE: "Skip", "Todo" and "Bail out!" literals are deferred to the parser
+  ]);
+
+  #isComment = false;
+  #source = null;
+  #line = 1;
+  #column = 0;
+  #escapeStack = [];
```

Comment:
Does this need to be an array? It looks like we are never reading its content, it might as well be a number.